### PR TITLE
Jmm/update for chicoma

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

Chicoma requires new and exciting changes to run. This PR incorporates those changes. These changes are:
1. Update to parthenon develop, due to the requirement of Kokkos 4. This necessitated some changes in the driver and boundary conditions to conform to Parthenon's current API. It also enabled me to thread through restart-able params. @mari2895 and @curiousmiah : this means that restarts should now work for self-gravity. @AstroBarker : I didn't thread the restart changes through the monte carlo control variables, but you can see from this diff how it's done. 
2. I added `nvcc_wrapper` to `scripts/bash`. This is a modified version of the wrapper provided by Kokkos but specifically set up to pass appropriate flags when building with the `nvhpc` environment. You can build as normally most places, but when building with `nvhpc`  you should set `-DCMAKE_CXX_COMPILER=path/to/phoebus/scripts/bash/nvcc_wrapper`. Note you need the full path.

## PR Checklist

- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by calling `scripts/bash/format.sh`.
- [x] Explain what you did.

### CODE REVIEW

File: src/monopole_gr/monopole_gr.cpp [hunk 1]
```diff
@@ -59,10 +59,11 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
   params.Add("force_static", force_static);
 
   // Only run the first n subcycles
+  auto restart = parthenon::Params::Mutability::Restart;
   int run_n_times = pin->GetOrAddInteger("monopole_gr", "run_n_times", -1);
   params.Add("run_n_times", run_n_times);
   int nth_call = 0;
-  params.Add("nth_call", nth_call, true); // mutable
+  params.Add("nth_call", nth_call, restart);
```
Comment (Yurlungur) [inline src/monopole_gr/monopole_gr.cpp:66]: @AstroBarker here's how you can make a param restart readable

File: src/monopole_gr/monopole_gr.cpp [hunk 2]
```diff
@@ -162,29 +163,28 @@ std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
...
+  params.Add("integration_volumes", integration_volumes, restart);
+  params.Add("integration_volumes_h", integration_vols_h, restart);
+  params.Add("hypersurface", hypersurface, restart);
+  params.Add("hypersurface_h", hypersurface_h, restart);
+  params.Add("lapse_h", alpha_h, restart);
+  params.Add("lapse", alpha, restart);
+  params.Add("lapse_last", alpha_last, restart);
+  auto mutable_param = parthenon::Params::Mutability::Mutable;
+  params.Add("matter_reducer", matter_reducer, mutable_param);
+  params.Add("volumes_reducer", volumes_reducer, mutable_param);
```
Comment (Yurlungur) [inline src/monopole_gr/monopole_gr.cpp:179]: And here's all the code setting the monopole solver to have appropriate restart settings

File: src/phoebus_boundaries/phoebus_boundaries.cpp [hunk 3]
```diff
@@ -32,6 +33,8 @@ using namespace parthenon::package::prelude;
 
 namespace Boundaries {
 
+parthenon::TopologicalElement CC = parthenon::TopologicalElement::CC;
```
Comment (Yurlungur) [inline src/phoebus_boundaries/phoebus_boundaries.cpp:36]: This code could probably be cleaned up now that @lroberts36 exposed generic boundary conditions but I went for the minimalist approach of just tweaking what we have

File: src/phoebus_driver.cpp [hunk 4]
```diff
@@ -130,27 +129,8 @@ void PhoebusDriver::PostInitializationCommunication() {
...
-    auto recv = tl.AddTask(
-        start_recv, parthenon::cell_centered_bvars::ReceiveBoundBufs<nonlocal>, md);
-    auto set = tl.AddTask(recv, parthenon::cell_centered_bvars::SetBounds<nonlocal>, md);
-
-    if (pmesh->multilevel) {
-      tl.AddTask(set | set_local, parthenon::cell_centered_bvars::RestrictGhostHalos, md,
-                 false);
-    }
+    auto boundary_tasks =
+        parthenon::AddBoundaryExchangeTasks(none, tl, md, pmesh->multilevel);
```
Comment (Yurlungur) [inline src/phoebus_driver.cpp:133]: @brryan and or @lroberts36 I'd like your eyes on the changes to the task list, in case I missed something.

Comment (brryan) [reply src/phoebus_driver.cpp:133]: This looks good to me (it's a very nice cleanup of the driver code actually) except I'm not familiar with `parthenon::AddBoundaryExchangeTasks` so maybe worth @lroberts36 taking a look as well.

File: src/phoebus_driver.cpp [hunk 5]
```diff
@@ -618,9 +576,9 @@ TaskCollection PhoebusDriver::RungeKuttaStage(const int stage) {
     auto &tl = async_region_3[i];
     auto &sc = pmb->meshblock_data.Get(stage_name[stage]);
 
-    auto prolongBound = tl.AddTask(none, parthenon::ProlongateBoundaries, sc);
+    // auto prolongBound = tl.AddTask(none, parthenon::ProlongateBoundaries, sc);
 
-    auto set_bc = tl.AddTask(prolongBound, parthenon::ApplyBoundaryConditions, sc);
+    auto set_bc = tl.AddTask(none, parthenon::ApplyBoundaryConditions, sc);
```
Comment (lroberts36) [inline src/phoebus_driver.cpp:581]: See above.

File: src/phoebus_driver.cpp [hunk 6]
```diff
@@ -159,9 +139,7 @@ void PhoebusDriver::PostInitializationCommunication() {
     auto &tl = async_region_2[i];
     auto &sc = pmb->meshblock_data.Get();
 
-    auto prolongBound = tl.AddTask(none, parthenon::ProlongateBoundaries, sc);
-
-    auto set_bc = tl.AddTask(prolongBound, parthenon::ApplyBoundaryConditions, sc);
+    auto set_bc = tl.AddTask(none, parthenon::ApplyBoundaryConditions, sc);
```
Comment (lroberts36) [inline src/phoebus_driver.cpp:142]: `parthenon::ApplyBoundaryConditions` (which I believe is filling boundaries on the fine level) is already taken care of by the tasks added by `AddBoundaryExchangeTasks` (although applying the boundary conditions twice shouldn't hurt anything I guess).

**Review by lroberts36 — APPROVED** [review]
LGTM, aside from my comments above.

**Comment by Yurlungur** [comment]
@brryan did you get to check this yet?

@mari2895 @AstroBarker please try it again should work now.

File: CMakeLists.txt [hunk 7]
```diff
@@ -168,12 +168,18 @@ set(CMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY ON CACHE BOOL "" FORCE)
 set(CMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY ON CACHE BOOL "" FORCE)
 
 # singularity eos
-message("Configuring singularity-eos")
-add_subdirectory(external/singularity-eos singularity-eos)
+message(STATUS "Patching mpark::variant to support GPUs")
+execute_process(COMMAND patch -N -s -V never
+  ${CMAKE_CURRENT_SOURCE_DIR}/../external/singularity-eos/utils/variant/include/mpark/variant.hpp
+  ${CMAKE_CURRENT_SOURCE_DIR}/../external/singularity-eos/utils/cuda_compatibility.patch
+)
```
Comment (Yurlungur) [inline CMakeLists.txt:175]: if we're ignoring singularity's `cmake` then we have to do this patch ourselves.

File: src/radiation/cooling_function.cpp [hunk 8]
```diff
@@ -169,6 +169,7 @@ TaskStatus CoolingFunctionCalculateFourForce(MeshBlockData<Real> *rc, const doub
   const bool do_liebendorfer = rad->Param<bool>("do_liebendorfer");
   const bool do_lightbulb = rad->Param<bool>("do_lightbulb");
   if (do_lightbulb) {
+#ifdef SPINER_USE_HDF
```
Comment (Yurlungur) [inline src/radiation/cooling_function.cpp:172]: required to build without HDF5

**Review by Yurlungur — COMMENTED** [review]
I have moved away from using the `singularity-eos` and `singularity-opac` cmake as for whatever reason I can't get HDF5 to cooperate between them all. But if I coordinate at the top level in phoebus, it works.

**Comment by AstroBarker** [comment]
Built and running on Chicoma with the latest changes

**Pull request merged**